vm/systemSmartContracts/stakingWaitingList.go

```diff
@@ -0,0 +1,1170 @@
+		s.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	if bytes.Equal(waitingElementKey, waitingListHead.FirstKey) {
+		s.eei.Finish([]byte(strconv.Itoa(1)))
+		return vmcommon.Ok
+	}
+	if bytes.Equal(waitingElementKey, waitingListHead.LastKey) {
+		s.eei.Finish([]byte(strconv.Itoa(int(waitingListHead.Length))))
+		return vmcommon.Ok
+	}
+
+	prevElement, err := s.getWaitingListElement(waitingListHead.FirstKey)
+	if err != nil {
+		s.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	index := uint32(2)
+	nextKey := make([]byte, len(waitingElementKey))
+	copy(nextKey, prevElement.NextKey)
+	for len(nextKey) != 0 && index <= waitingListHead.Length {
+		if bytes.Equal(nextKey, waitingElementKey) {
+			s.eei.Finish([]byte(strconv.Itoa(int(index))))
+			return vmcommon.Ok
+		}
+
+		prevElement, err = s.getWaitingListElement(nextKey)
+		if err != nil {
+			s.eei.AddReturnMessage(err.Error())
+			return vmcommon.UserError
+		}
+
+		if len(prevElement.NextKey) == 0 {
+			break
+		}
+		index++
+		copy(nextKey, prevElement.NextKey)
+	}
+
+	s.eei.AddReturnMessage("element in waiting list not found")
+	return vmcommon.UserError
+}
+
+func (s *stakingSC) getWaitingListSize(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if s.flagStakingV4.IsSet() {
+		s.eei.AddReturnMessage(vm.ErrWaitingListDisabled.Error())
+		return vmcommon.UserError
+	}
+
+	if args.CallValue.Cmp(zero) != 0 {
+		s.eei.AddReturnMessage(vm.TransactionValueMustBeZero)
+		return vmcommon.UserError
+	}
+
+	err := s.eei.UseGas(s.gasCost.MetaChainSystemSCsCost.Get)
+	if err != nil {
+		s.eei.AddReturnMessage("insufficient gas")
+		return vmcommon.OutOfGas
+	}
+
+	waitingListHead, err := s.getWaitingListHead()
+	if err != nil {
+		s.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	s.eei.Finish([]byte(strconv.Itoa(int(waitingListHead.Length))))
+	return vmcommon.Ok
+}
+
+func (s *stakingSC) getWaitingListRegisterNonceAndRewardAddress(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !bytes.Equal(args.CallerAddr, s.stakeAccessAddr) {
+		s.eei.AddReturnMessage("this is only a view function")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) != 0 {
+		s.eei.AddReturnMessage("number of arguments must be equal to 0")
+		return vmcommon.UserError
+	}
+
+	waitingListData, err := s.getFirstElementsFromWaitingList(math.MaxUint32)
+	if err != nil {
+		s.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if len(waitingListData.stakedDataList) == 0 {
+		s.eei.AddReturnMessage("no one in waitingList")
+		return vmcommon.UserError
+	}
+
+	for index, stakedData := range waitingListData.stakedDataList {
+		s.eei.Finish(waitingListData.blsKeys[index])
+		s.eei.Finish(stakedData.RewardAddress)
+		s.eei.Finish(big.NewInt(int64(stakedData.RegisterNonce)).Bytes())
+	}
+
+	return vmcommon.Ok
+}
+
+func (s *stakingSC) getTotalNumberOfRegisteredNodes(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```

Comment:
this is not old code - this can be moved back.

Reply:
Right, this is not old, it just used the waiting list, but it is not _legacy_. Moved it back